### PR TITLE
add support for yaml test source files

### PR DESCRIPTION
**Pull request by winsvega**
boost getFile by extention mask

this is a step by step PR for Yoichi.

### CODE REVIEW

File: test/tools/libtesteth/TestHelper.cpp [hunk 1]
```diff
@@ -242,21 +243,31 @@ void replaceLLLinState(json_spirit::mObject& _o)
 	}
 }
 
-std::vector<boost::filesystem::path> getJsonFiles(boost::filesystem::path const& _dirPath, std::string const& _particularFile)
+std::vector<boost::filesystem::path> getFiles(boost::filesystem::path const& _dirPath, string const& _extentionMask, std::string const& _particularFile)
```
Comment (axic) [inline test/tools/libtesteth/TestHelper.cpp]: Why not just use a `vector` for the extensions?

Comment (chfast) [reply test/tools/libtesteth/TestHelper.cpp]: Just remove this param and hardcode expected extensions in getFiles. That would save us some parsing magic.

But set is also good. You can do:
```cpp
#include <set>

auto f(std::set<const char*> s)
{
  return s.size();
}

int main()
{
  return f({".json", ".yml"});
}
```

File: test/tools/libtesteth/TestHelper.cpp [hunk 2]
```diff
@@ -242,21 +243,31 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
 {
 	vector<boost::filesystem::path> jsonFiles;
+	vector<string> extentions;
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: If you want a `vector` here use it as an argument. Another alternative to `std::set`.

File: test/tools/libtesteth/TestHelper.cpp [hunk 3]
```diff
@@ -242,21 +243,31 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+	boost::split(extentions, _extentionMask, boost::is_any_of("|"));
 	if (!_particularFile.empty())
 	{
-		boost::filesystem::path file = _dirPath / (_particularFile + ".json");
-		if (boost::filesystem::exists(file))
-			jsonFiles.push_back(file);
+		for(auto const& ext: extentions)
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: If `extensions` contains duplicates you will add the same file many times.

File: test/tools/libtesteth/TestSuite.cpp [hunk 4]
```diff
@@ -122,26 +122,29 @@ void TestSuite::runAllTestsInFolder(string const& _testFolder) const
 {
 	// check that destination folder test files has according Filler file in src folder
 	string const filter = test::Options::get().singleTestName.empty() ? string() : test::Options::get().singleTestName;
-	vector<fs::path> const compiledFiles = test::getJsonFiles(getFullPath(_testFolder), filter);
+	vector<fs::path> const compiledFiles = test::getFiles(getFullPath(_testFolder), ".json|.yaml" ,filter);
```
Comment (chfast) [inline test/tools/libtesteth/TestSuite.cpp]: Use `.yml` not `.yaml`. See `.travis.yml`, `circle.yml`, etc.

File: test/tools/libtesteth/TestSuite.cpp [hunk 5]
```diff
@@ -122,26 +122,29 @@ void TestSuite::runAllTestsInFolder(string const& _testFolder) const
...
 	for (auto const& file: compiledFiles)
 	{
 		fs::path const expectedFillerName = getFullPathFiller(_testFolder) / fs::path(file.stem().string() + c_fillerPostf + ".json");
+		fs::path const expectedFillerName2 = getFullPathFiller(_testFolder) / fs::path(file.stem().string() + c_fillerPostf + ".yaml");
```
Comment (chfast) [inline test/tools/libtesteth/TestSuite.cpp]: So you have the magic `".json|.yaml"` but here you hard-coded the extensions?

File: test/CMakeLists.txt [hunk 6]
```diff
@@ -50,9 +50,12 @@ foreach(file ${sources})
 	endforeach(test_raw)
 endforeach(file)
 
+hunter_add_package(yaml-cpp)
```
[the remark this answers is not in the document]
Comment (winsvega) [reply test/CMakeLists.txt]: there is no yaml code yet. I want to do it in a separate commit

File: test/tools/libtesteth/TestHelper.cpp (hunk 3 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: true

File: test/tools/libtesteth/TestSuite.cpp (hunk 5 above)
Comment (winsvega) [reply test/tools/libtesteth/TestSuite.cpp]: the function could be used for more general perpuses 
hardcode is bad.

File: test/tools/libtesteth/TestHelper.cpp [hunk 7]
```diff
@@ -242,21 +243,51 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+		}
 	}
 	else
 	{
 		using Bdit = boost::filesystem::directory_iterator;
 		for (Bdit it(_dirPath); it != Bdit(); ++it)
-			if (boost::filesystem::is_regular_file(it->path()) && it->path().extension() == ".json")
-					jsonFiles.push_back(it->path());
+		{
+			for(auto const& ext: extentions)
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: This code looks similar to the loop above. Can you refactor it to have single loop of extensions?

File: test/tools/libtesteth/TestSuite.cpp [hunk 8]
```diff
@@ -122,26 +122,29 @@ void TestSuite::runAllTestsInFolder(string const& _testFolder) const
 {
 	// check that destination folder test files has according Filler file in src folder
 	string const filter = test::Options::get().singleTestName.empty() ? string() : test::Options::get().singleTestName;
-	vector<fs::path> const compiledFiles = test::getJsonFiles(getFullPath(_testFolder), filter);
+	vector<fs::path> const compiledFiles = test::getFiles(getFullPath(_testFolder), {".json",".yaml"} ,filter);
 	for (auto const& file: compiledFiles)
 	{
 		fs::path const expectedFillerName = getFullPathFiller(_testFolder) / fs::path(file.stem().string() + c_fillerPostf + ".json");
+		fs::path const expectedFillerName2 = getFullPathFiller(_testFolder) / fs::path(file.stem().string() + c_fillerPostf + ".yaml");
```
Comment (chfast) [inline test/tools/libtesteth/TestSuite.cpp]: If you have "generic" function getFiles(), why do you have hand-crafted errors here?

File: test/tools/libtesteth/TestSuite.cpp [hunk 9]
```diff
@@ -122,26 +122,29 @@ void TestSuite::runAllTestsInFolder(string const& _testFolder) const
 {
 	// check that destination folder test files has according Filler file in src folder
 	string const filter = test::Options::get().singleTestName.empty() ? string() : test::Options::get().singleTestName;
-	vector<fs::path> const compiledFiles = test::getJsonFiles(getFullPath(_testFolder), filter);
+	vector<fs::path> const compiledFiles = test::getFiles(getFullPath(_testFolder), {".json",".yaml"} ,filter);
```
Comment (chfast) [inline test/tools/libtesteth/TestSuite.cpp]: 1. Wrong formatting.
2. `".yml"` not `".yaml"`.

File: test/tools/libtesteth/TestSuite.cpp [hunk 10]
```diff
@@ -122,26 +122,29 @@ void TestSuite::runAllTestsInFolder(string const& _testFolder) const
...
+			if (fs::exists(expectedFillerName2))
+				checkFillerHash(file, expectedFillerName2);
 			if (fs::exists(expectedCopierName))
 				checkFillerHash(file, expectedCopierName);
 		}
 	}
 
 	// run all tests
-	vector<fs::path> const files = test::getJsonFiles(getFullPathFiller(_testFolder), filter.empty() ? filter : filter + "Filler");
+	vector<fs::path> const files = test::getFiles(getFullPathFiller(_testFolder), {".json",".yaml"}, filter.empty() ? filter : filter + "Filler");
```
Comment (chfast) [inline test/tools/libtesteth/TestSuite.cpp]: 1. Wrong formatting.
2. `".yml"` not `".yaml"`.

File: test/tools/libtesteth/TestSuite.cpp (hunk 8 above)
Comment (winsvega) [reply test/tools/libtesteth/TestSuite.cpp]: this is a check for the filler name. each test must have only one filler file. 
A filler could be  Filler.json, Filler.yml and Copier.json

File: test/tools/libtesteth/TestSuite.cpp (hunk 10 above)
Comment (winsvega) [reply test/tools/libtesteth/TestSuite.cpp]: what's wrong for 1) ?

File: test/tools/libtesteth/TestHelper.cpp (hunk 7 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: I think this two for loop's could be moved upwards and combined.

File: test/tools/libtesteth/TestSuite.cpp (hunk 10 above)
Comment (chfast) [reply test/tools/libtesteth/TestSuite.cpp]: `","`

File: test/tools/libtesteth/TestHelper.cpp [hunk 11]
```diff
@@ -242,21 +243,51 @@ void replaceLLLinState(json_spirit::mObject& _o)
 	}
 }
 
-std::vector<boost::filesystem::path> getJsonFiles(boost::filesystem::path const& _dirPath, std::string const& _particularFile)
+template<class T>
+std::set<T> removeRepeatElementsFromSet(std::set<T> const& _set)
```
[the remark this answers is not in the document]
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: done

File: test/unittests/libtesteth/testHelper.cpp [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+BOOST_FIXTURE_TEST_SUITE(TestHelperTestSuite, TestOutputHelperFixture)
+
+BOOST_AUTO_TEST_CASE(removeElementsFromSetTest_doubleElements)
+{
+	set<string> aSet;
+	aSet.insert("value1");
+	aSet.insert("value2");
+	aSet.insert("value1");
+	aSet = test::removeRepeatElementsFromSet(aSet);
+	BOOST_REQUIRE(aSet.size() == 2);
```
Comment (chfast) [inline test/unittests/libtesteth/testHelper.cpp]: Can you add another `BOOST_REQUIRE` before `test::removeRepeatElementsFromSet()`

**Comment by winsvega** [comment]
ok now we could convert some existing tests into yaml and see what happens.

File: test/CMakeLists.txt [hunk 13]
```diff
@@ -68,6 +71,7 @@ eth_add_test(ClientBase
 	ARGS --eth_testfile=BlockTests/bcValidBlockTest --eth_threads=10
 )
 
+
```
Comment (pirapira) [inline test/CMakeLists.txt]: ?

File: test/tools/libtesteth/ImportTest.cpp [hunk 14]
```diff
@@ -28,6 +28,7 @@
 #include <test/tools/jsontests/BlockChainTests.h>
 
 #include <boost/filesystem/path.hpp>
+#include <yaml-cpp/yaml.h>
```
Comment (pirapira) [inline test/tools/libtesteth/ImportTest.cpp]: I don't think this include is necessary in this commit.

File: test/tools/libtesteth/ImportTest.cpp [hunk 15]
```diff
@@ -524,13 +524,21 @@ void ImportTest::parseJsonStrValueIntoVector(json_spirit::mValue const& _json, v
...
+
 	if (_json.type() == json_spirit::array_type)
 	{
 		for (auto const& val: _json.get_array())
-			_out.push_back(val.get_int());
+			converter(val);
 	}
 	else
-		_out.push_back(_json.get_int());
+		converter(_json);
```
Comment (pirapira) [inline test/tools/libtesteth/ImportTest.cpp]: I don't think it's a good idea to allow both int and str here. Let's just pick one of these and stick to that.  Maybe an `int`.

File: test/tools/libtesteth/TestSuite.cpp [hunk 16]
```diff
@@ -150,7 +150,7 @@ void TestSuite::runAllTestsInFolder(string const& _testFolder) const
 	for (auto const& file: files)
 	{
 		testOutput.showProgress();
-		testOutput.setCurrentTestFileName(file.filename().string());
+		testOutput.setCurrentTestFileName(file);
```
Comment (pirapira) [inline test/tools/libtesteth/TestSuite.cpp]: This change is not related to other changes in the same commit.

File: test/tools/libtesteth/ImportTest.cpp (hunk 15 above)
Comment (winsvega) [reply test/tools/libtesteth/ImportTest.cpp]: well. yaml does not know which type of the field it's parsing. and assume that it is string. which is true for all our json fields. but this one. we already have all tests with this fields as int. 
but after yaml converts this into json it becomes string. so either I force this field back to string by adding a new function or we keep this change which allows integer array in a string form.

File: test/tools/libtesteth/TestSuite.cpp (hunk 16 above)
Comment (winsvega) [reply test/tools/libtesteth/TestSuite.cpp]: is there a way to remove a particular change from a commit?

File: test/tools/libtesteth/ImportTest.cpp (hunk 15 above)
Comment (pirapira) [reply test/tools/libtesteth/ImportTest.cpp]: You can try this https://github.com/ethereum/tests/issues/369#issuecomment-352452404

File: test/unittests/libtesteth/testHelper.cpp (hunk 12 above)
Comment (chfast) [reply test/unittests/libtesteth/testHelper.cpp]: This test does not test anything.

**Comment by winsvega** [comment]
it tests that elements not removed.

**Comment by chfast** [comment]
How do you know if `aSet.size()` is not 2 **before** `test::removeRepeatElementsFromSet(aSet)`?

**Comment by winsvega** [comment]
I rely on std class implementation already being tested

**Comment by chfast** [comment]
You are completely wrong.

> std::set is an associative container that contains a sorted set of **unique** objects of type Key.

You wasted a lot of your and our time to write a code that is totally useless. I hoped you can spot the mistake so suggested to write an unit test. But nope.

**Comment by winsvega** [comment]
so you deliberatly misleaded me?

**Comment by pirapira** [comment]
@chfast it takes time when you try to make someone realize their own mistakes.

@winsvega it's quite common that people say "write a test" when they spot a mistake.

File: test/tools/libtesteth/TestHelper.cpp [hunk 17]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+
+	if (_node.IsSequence())
+	{
+		json_spirit::mArray jArray;
+		for (std::size_t i = 0; i < _node.size(); i++)
+			jArray.push_back(convertYamlNodeToJson(_node[i]));
+		return jArray;
+	}
+
+	return json_spirit::mValue();
```
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: I prefer an error here, rather than silently ignoring an unknown.

File: test/tools/libtesteth/TestSuite.cpp (hunk 16 above)
Comment (winsvega) [reply test/tools/libtesteth/TestSuite.cpp]: changed

File: test/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (winsvega) [reply test/CMakeLists.txt]: changed

File: test/tools/libtesteth/TestHelper.cpp (hunk 17 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: ok. will add boost error here

File: test/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (chfast) [reply test/CMakeLists.txt]: What changed?

File: test/tools/jsontests/BlockChainTests.cpp [hunk 18]
```diff
@@ -53,11 +53,11 @@ json_spirit::mValue BlockchainTestSuite::doTests(json_spirit::mValue const& _inp
 			continue;
 
 		BOOST_REQUIRE_MESSAGE(inputTest.count("genesisBlockHeader"),
-			"\"genesisBlockHeader\" field is not found. filename: " + TestOutputHelper::get().testFileName() +
+			"\"genesisBlockHeader\" field is not found. filename: " + TestOutputHelper::get().testFileName().string() +
```
Comment (chfast) [inline test/tools/jsontests/BlockChainTests.cpp]: Why suddenly we have `.string()` everywhere?

File: test/tools/libtesteth/TestHelper.cpp [hunk 19]
```diff
@@ -28,7 +28,9 @@
 
 #include <boost/algorithm/string/trim.hpp>
 #include <boost/filesystem/path.hpp>
+#include <set>
 #include <string>
+#include <yaml-cpp/yaml.h>
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: The order of #includes is wrong. yaml-cpp must be included before boost.

File: test/tools/libtesteth/TestHelper.cpp [hunk 20]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
 	}
 }
 
-std::vector<boost::filesystem::path> getJsonFiles(boost::filesystem::path const& _dirPath, std::string const& _particularFile)
+std::vector<boost::filesystem::path> getFiles(boost::filesystem::path const& _dirPath, std::set<string> const _extentionMask, std::string const& _particularFile)
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Inconsistent namespace usage. Alias `boost::filesystem` as `fs` and use `std::string`.

File: test/tools/libtesteth/TestHelper.cpp [hunk 21]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
-			jsonFiles.push_back(file);
+		if (!_particularFile.empty())
+		{
+			boost::filesystem::path file = _dirPath / (_particularFile + ext);
+			if (boost::filesystem::exists(file))
+				files.push_back(file);
+		}
+		else
+		{
+			using Bdit = boost::filesystem::directory_iterator;
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Stupid name.

File: test/tools/libtesteth/TestHelper.cpp [hunk 22]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+			}
+		}
 	}
-	else
+	return files;
+}
+
+json_spirit::mValue convertYamlNodeToJson(YAML::Node _node)
+{
+	if (_node.IsNull())
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Why do you map `null` to `{}`?

File: test/tools/libtesteth/TestHelper.cpp [hunk 23]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+
+	if (_node.IsScalar())
 	{
-		using Bdit = boost::filesystem::directory_iterator;
-		for (Bdit it(_dirPath); it != Bdit(); ++it)
-			if (boost::filesystem::is_regular_file(it->path()) && it->path().extension() == ".json")
-					jsonFiles.push_back(it->path());
+		json_spirit::mValue value;
+		if (_node.Tag() == "tag:yaml.org,2002:int")
+			value = _node.as<int>();
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: You can simplify it by `if (_node.Tag() == "tag:yaml.org,2002:int") return _node.as<int>();`.

File: test/tools/libtesteth/TestHelper.cpp [hunk 24]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+			value = _node.as<std::string>();
+		return value;
 	}
-	return jsonFiles;
+
+	if (_node.IsMap())
+	{
+		json_spirit::mObject jObject;
+		for (auto const& i : _node)
+			jObject[i.first.as<std::string>()] = convertYamlNodeToJson(i.second);
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Use `std::map::emplace`.

File: test/tools/libtesteth/TestHelper.cpp [hunk 25]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+		return jObject;
+	}
+
+	if (_node.IsSequence())
+	{
+		json_spirit::mArray jArray;
+		for (std::size_t i = 0; i < _node.size(); i++)
+			jArray.push_back(convertYamlNodeToJson(_node[i]));
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Use `emplace_back()`.

File: test/tools/libtesteth/TestHelper.cpp [hunk 26]
```diff
@@ -242,23 +244,69 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+		return jArray;
+	}
+
+	return json_spirit::mValue();
+}
+
+// this function is here so not to include YAML into other files
```
Comment (chfast) [inline test/tools/libtesteth/TestHelper.cpp]: Function comments should start with `///` followed by proper sentence.

File: test/tools/jsontests/BlockChainTests.cpp (hunk 18 above)
Comment (winsvega) [reply test/tools/jsontests/BlockChainTests.cpp]: changed to boost path

File: test/tools/libtesteth/TestHelper.cpp (hunk 22 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: empty json fields

Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: fixed

File: test/tools/libtesteth/TestHelper.cpp [hunk 27]
```diff
@@ -242,26 +244,71 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+
+	if (_node.IsSequence())
+	{
+		json_spirit::mArray jArray;
+		for (size_t i = 0; i < _node.size(); i++)
+			jArray.emplace_back(convertYamlNodeToJson(_node[i]));
+		return jArray;
+	}
+
+	BOOST_ERROR("Error parsing YAML node. Element type note defined!");
```
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: `not defined` instead of `note defined`.

File: test/CMakeLists.txt (hunk 6 above)
[the remark this answers is not in the document]
Comment (winsvega) [reply test/CMakeLists.txt]: yaml code and this line are in the same PR

File: test/tools/libtesteth/TestSuite.cpp (hunk 16 above)
Comment (chfast) [reply test/tools/libtesteth/TestSuite.cpp]: Why do you use full path now? The name of the function is confusing.

Comment (winsvega) [reply test/tools/libtesteth/TestSuite.cpp]: because file is not could be Filler.json and Filler.yml 
boost path provide interface to get filename and extention

**Comment by winsvega** [comment]
because file is not could be Filler.json and Filler.yml
boost path provide interface to get filename and extention

File: test/tools/libtesteth/TestSuite.cpp (hunk 16 above)
Comment (chfast) [reply test/tools/libtesteth/TestSuite.cpp]: So you should rename the function to `setCurrentTestFile()`.

File: test/tools/libtesteth/TestHelper.cpp [hunk 28]
```diff
@@ -271,45 +271,44 @@ std::vector<boost::filesystem::path> getFiles(boost::filesystem::path const& _di
...
 		return jArray;
 	}
 
+	BOOST_ERROR("Error parsing YAML node. Element type note defined!");
 	return json_spirit::mValue();
 }
 
-// this function is here so not to include YAML into other files
-json_spirit::mValue parseYamlToJson(std::string const& _string)
+/// this function is here so not to include YAML into other files
```
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: I don't understand this phrase: "Include YAML into other files".

* including YAML in JSON files?  Why would that happen?
* including YAML in CPP sources?  Maybe not.

File: test/tools/libtesteth/TestHelper.cpp [hunk 29]
```diff
@@ -242,25 +243,32 @@ void replaceLLLinState(json_spirit::mObject& _o)
...
+					files.push_back(it->path());
+			}
+		}
 	}
-	return jsonFiles;
+	return files;
 }
 
+
+
```
Comment (pirapira) [inline test/tools/libtesteth/TestHelper.cpp]: ?

File: test/tools/libtesteth/TestHelper.cpp (hunk 28 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: so not to include yaml headers in other cpp files in project cpp-ethereum/test

File: test/tools/libtesteth/TestHelper.cpp (hunk 29 above)
Comment (winsvega) [reply test/tools/libtesteth/TestHelper.cpp]: this is already fixed
